internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,668 @@
+		for diff < base {
+			t, err := iterProducers[1].iter.Next(ctx)
+			if err != nil {
+				if storage.IterIsDoneOrCancelled(err) {
+					iterProducers[1].iter.Stop()
+					iterProducers[1].iter = nil
+					break
+				}
+				// this would be weird
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+			diff = t.GetObject()
+		}
+	}
+
+	// drain the base
+	if len(iterProducers) == 1 && iterProducers[0].idx == 0 {
+		for len(iterProducers) == 1 && iterProducers[0].iter != nil {
+			for {
+				t, err := iterProducers[0].iter.Next(ctx)
+				if err != nil {
+					if storage.IterIsDoneOrCancelled(err) {
+						iterProducers[0].iter.Stop()
+						iterProducers[0].iter = nil
+						break
+					}
+					concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+					return
+				}
+				batch = append(batch, t)
+			}
+			var err error
+			iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+			if err != nil {
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+		}
+	}
+}
+
+func (c *LocalChecker) fastPathOperationSetup(ctx context.Context, req *ResolveCheckRequest, op setOperatorType, children ...*openfgav1.Userset) (chan *iteratorMsg, error) {
+	iterProducers := make([]*iteratorProducer, 0, len(children))
+	for idx, child := range children {
+		producerChan, err := c.fastPathRewrite(ctx, req, child)
+		if err != nil {
+			return nil, err
+		}
+		iterProducers = append(iterProducers, &iteratorProducer{idx: idx, producer: producerChan})
+	}
+	var resolver fastPathSetHandler
+	switch op {
+	case unionSetOperator:
+		resolver = fastPathUnion
+	case intersectionSetOperator:
+		resolver = fastPathIntersection
+	case exclusionSetOperator:
+		resolver = fastPathDifference
+	default:
+		return nil, ErrUnknownSetOperator
+	}
+	resultChan := make(chan *iteratorMsg, len(children))
+	go resolver(ctx, iterProducers, resultChan)
+	return resultChan, nil
+}
+
+func (c *LocalChecker) fastPathRewrite(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+	rewrite *openfgav1.Userset,
+) (chan *iteratorMsg, error) {
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgav1.Userset_This:
+		return c.fastPathDirect(ctx, req)
+	case *openfgav1.Userset_ComputedUserset:
+		return c.fastPathComputed(ctx, req, rewrite)
+	case *openfgav1.Userset_Union:
+		return c.fastPathOperationSetup(ctx, req, unionSetOperator, rw.Union.GetChild()...)
+	case *openfgav1.Userset_Intersection:
+		return c.fastPathOperationSetup(ctx, req, intersectionSetOperator, rw.Intersection.GetChild()...)
+	case *openfgav1.Userset_Difference:
+		return c.fastPathOperationSetup(ctx, req, exclusionSetOperator, rw.Difference.GetBase(), rw.Difference.GetSubtract())
+	default:
+		return nil, ErrUnknownSetOperator
+	}
+}
+
+func (c *LocalChecker) resolveFastPath(ctx context.Context, leftChan <-chan *iteratorMsg, iter TupleMapper) (*ResolveCheckResponse, error) {
+	rightOpen := true
```

Comment:
can you add some comment to explain why is left vs. right?  Hard to picture in my head what does that really mean.